### ISSUE TITLE
Contrastive loss with non-binary labels

### ISSUE THREAD

**Issue by Garrafao** [UKPLab/sentence-transformers#3382]
Contrastive loss allows for training with non-binary labels under the hood, i.e., it doesn't throw an error if you provide labels like 0.1 or 0.6. Maybe it would be good to catch non-binary labels in the function with an assertion statement to avoid misapplication?

**Comment by tomaarsen** [comment]
Hello!

This indeed likely sounds like it's somewhat unintentional, but I have a suspicion that it won't actually be disastrous for the training. Perhaps it's okay to let users do this, even if accidental?

- Tom Aarsen

**Comment by Garrafao** [comment]
Hi Tom,

you're probably right as we have unintentionally tried that and results were good. But, maybe a warning would be good at least?

An attempt to explain why it works:
Assume you give cosine-like labels as input to the [contrastive loss](https://github.com/UKPLab/sentence-transformers/blob/163258f16cc0fa75639fd6e2b58bdc42921178f0/sentence_transformers/losses/ContrastiveLoss.py#L101). That's the definition:

`losses = 0.5 * (labels.float() * distances.pow(2) + (1 - labels).float() * F.relu(self.margin - distances).pow(2))`

For one pair, setting `self.margin=0.5` (default value), and ignoring the scaling factor `0.5` this is simplified

`loss = label * distance^2 + (1 - label) * max(0,0.5 - distance^2)`

Let's assume `label=0`, then

`loss = max(0,0.5 - distance^2)`

So, the loss is minimized at `distance>=sqrt(0.5)`

Let's assume `label=0.1`, then

`loss = 0.1 * distance^2 + 0.9 * max(0,0.5 - distance^2)`

This loss is also minimized at `distance=sqrt(0.5)` and so is any `label<0.5`, check [here](https://www.wolframalpha.com/input?i=minimum+calculator&assumption=%7B%22F%22%2C+%22GlobalMinimizeCalculator%22%2C+%22curvefunction%22%7D+-%3E%22+0.1+*+x%5E2+%2B+0.9+*+max%280%2C0.5+-+x%5E2%29%22).

Similarly, for `label>=0.5` the loss is minimized at `distance=0.0`. If this is true, then training contrastive loss with dense labels is roughly is the same as mapping each `label>=0.5` to `1` and the rest to `0` (basically binarizing dense labels) and training contrastive loss with these.

Any thoughts?

**Comment by tomaarsen** [comment]
Very nice! Then it seems like non-binarized labels automatically get rounded/binarized. That seems quite helpful for users accidentally using cosine-like labels. I'm okay with adding a warning, but I think it's not very important. 

- Tom Aarsen
